File: 00020. Valid Parentheses/solution.go
/**
 * Problem: 20. Valid Parentheses
 *
 * Difficulty: Easy
 *
 * Language: Go
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func isValid(s string) bool {
	// Use a stack to track opening brackets
	stack := make([]rune, 0)
	
	// Map closing brackets to their corresponding opening brackets
	bracketMap := map[rune]rune{
		')': '(',
		']': '[',
		'}': '{',
	}

	// Process each character in the string
	for _, char := range s {
		switch char {
			case '(', '[', '{':
				// For opening brackets, push to stack
				stack = append(stack, char)
			case ')', ']', '}':
				// For closing brackets, check if matching opening bracket is at top of stack
				if len(stack) == 0 || stack[len(stack)-1] != bracketMap[char] {
					return false
				}
				// Remove the matched opening bracket from stack
				stack = stack[:len(stack)-1]
		}
	}

	// Valid if all brackets are matched (stack empty)
	return len(stack) == 0
}

File: 00020. Valid Parentheses/solution.js
/**
 * Problem: 20. Valid Parentheses
 *
 * Difficulty: Easy
 *
 * Language: JavaScript
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

/**
 * Validates string containing parentheses
 *
 * @param {string} s - Input string with brackets
 *
 * @returns {boolean} - True if brackets are valid
 */
const isValid = (s) => {
  // Create an empty stack to track opening brackets
  const stack = []
  // Map closing brackets to their corresponding opening brackets
  const pairs = { ')': '(', ']': '[', '}': '{' }

  // Iterate through each character in the string
  for (const char of s) {
    // If character is an opening bracket, push to stack
    if (!pairs[char]) stack.push(char)
    // If character is a closing bracket, check if it matches with the last opening bracket
    else if (stack.pop() !== pairs[char]) return false
  }

  // Valid only if all brackets were properly closed (stack is empty)
  return stack.length === 0
}
